src/storage/src/source/mod.rs

```diff
@@ -986,13 +1041,31 @@ where
 
             let mut timestamp_interval = tokio::time::interval(timestamp_frequency);
             timestamp_interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+            let mut pending_untimestamped_messages = vec![];
             let mut pending_messages = vec![];
             loop {
                 tokio::select! {
                     // N.B. This branch is cancel-safe because `next` only borrows the underlying stream.
                     item = source_stream.next() => {
                         match item {
-                            Some(Ok(message)) => pending_messages.push(Ok(message)),
+                            Some(Ok(message)) => match message {
+                                // Note that this
+                                // 1. Requires that sources that produce `InProgress` messages
+                                //    ALWAYS produce a `Finalized` for the final message.
+                                // 2. Requires that sources that produce `InProgress` messages
+                                //    NEVER produces messages at offsets below the most recent
+                                //    `Finalized` message.
+                                // 3. Buffers EVERY message associated with a single offset. This
+                                //    can be improved, tracked in
+                                //    <https://github.com/MaterializeInc/materialize/issues/12557>
+                                SourceMessageType::Finalized(message) => {
+                                    pending_messages.extend(pending_untimestamped_messages.drain(..));
```

Comment:
should this be keyed by partition id?  Offset is a property of the partition (perhaps this doesn't matter at the moment because the set of partitioned sources and the set of sources that will return `InProgress` may be non-overalpping

Reply:
oh yes, it should be a map with a partition key!